internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+		ctx = propagateOtelHeaders(ctx, conn.Spec().IsClient, conn.RequestHeader())
+		requestStartTime := time.Now()
+		name := strings.TrimLeft(conn.Spec().Procedure, "/")
+
+		attributes := getAttributes(ctx, "grpc")
+		traceOpts := []trace.SpanStartOption{
+			trace.WithAttributes(attributes...),
+			trace.WithSpanKind(trace.SpanKindServer),
+		}
+		tracer := otel.GetTracerProvider().Tracer(conn.Spec().Procedure)
+		ctx, span := tracer.Start(ctx, name, traceOpts...)
+		defer span.End()
+
+		instruments := getInstruments(conn.Spec().IsClient)
+		state := &streamingState{
+			spec:        conn.Spec(),
+			protocol:    conn.Peer().Protocol,
+			attributes:  attributes,
+			receiveSize: instruments.responseSize,
+			sendSize:    instruments.requestSize,
+		}
+		streamingHandler := &streamingHandlerInterceptor{
+			StreamingHandlerConn: conn,
+			receive: func(msg any, conn connect.StreamingHandlerConn) error {
+				return state.receive(ctx, msg, conn)
+			},
+			send: func(msg any, conn connect.StreamingHandlerConn) error {
+				return state.send(ctx, msg, conn)
+			},
+		}
+		err := next(ctx, streamingHandler)
+		state.attributes = append(
+			state.attributes,
+			statusCodeAttribute(conn.Peer().Protocol, err))
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+		}
+		span.SetAttributes(state.attributes...)
+		instruments.requestsPerRPC.Record(
+			ctx,
+			state.receivedCounter,
+			metric.WithAttributes(state.attributes...))
+		instruments.responsesPerRPC.Record(
+			ctx,
+			state.sentCounter,
+			metric.WithAttributes(state.attributes...))
+		instruments.duration.Record(ctx,
+			time.Since(requestStartTime).Milliseconds(),
+			metric.WithAttributes(state.attributes...))
+		return err
+	}
+}
+
+func propagateOtelHeaders(ctx context.Context, isClient bool, header http.Header) context.Context {
+	if isClient {
+		otel.GetTextMapPropagator().Inject(ctx, propagation.HeaderCarrier(header))
+	} else {
+		ctx = otel.GetTextMapPropagator().Extract(ctx, propagation.HeaderCarrier(header))
+	}
+	return ctx
+}
+
+type instrumentation struct {
+	duration        metric.Int64Histogram
+	requestSize     metric.Int64Histogram
+	responseSize    metric.Int64Histogram
+	requestsPerRPC  metric.Int64Histogram
+	responsesPerRPC metric.Int64Histogram
+}
+
+func createInstruments(meter metric.Meter) instrumentation {
+	duration, err := meter.Int64Histogram("duration", metric.WithUnit("ms"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create duration metric: %w", err))
+	}
+	requestSize, err := meter.Int64Histogram("request.size", metric.WithUnit("By"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create request size metric: %w", err))
+	}
+	responseSize, err := meter.Int64Histogram("response.size", metric.WithUnit("By"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create response size metric: %w", err))
+	}
+	requestsPerRPC, err := meter.Int64Histogram("requests_per_rpc", metric.WithUnit("1"))
```

Comment:
Did you find any docs on how unit is supposed to be set for counts? I wasn't sure if we should just leave it out. "1" makes sense intuitively but I'm mostly wondering if this is how we're supposed to do it for all the count metrics

Reply:
Yup, docs say that for counts the unit should be `"1"` 🤷

Reply:
That's so interesting. Seems like some sort of shorthand, but it is what they recommend :)